### ISSUE TITLE
Can't unhold orders

### ISSUE THREAD

**Issue by James-Moon-Audio** [signifyd/magento2#7]
Magento 2.19

If the Signifyd extension is installed we are unable to remove orders from hold status if the Signifyd Guarantee Status is N/A.

**Comment by ebanolopes** [comment]
Hi, @James-Moon-Audio, I'll take a look at this issue and return to you ASAP. Can you please check the Signifyd extension version installed on your Magento?

**Comment by James-Moon-Audio** [comment]
2.0.8

**Comment by ebanolopes** [comment]
I've performed a test here with the same Magento and Signifyd extension versions and haven't been able to reproduce the same situation. I've also done a quick check the extension source code and found nothing that could interfere at the release of the hold status of the order.

Can you please let me know how the extension is configured (please, do not post API key here, it is public):
- The extension is enabled?
- Is the API key correctly configured and cases been sent to Signifyd?
- How are configured "Action on Guarantee Approve" and "Action on Guarantee Decline"

If you can clarify to me how did you get the conclusion that it is the Signifyd extension that is preventing the order from been removed from the hold status, that could be helpful.

Also, please check Apache/Nginx and PHP logs to check if there is any error generated at the moment you try to unhold the order. If so, please, post/attach the erros here to me.

Thanks!

**Comment by James-Moon-Audio** [comment]
-The extension is enabled.
 - The API key is correctly configured.  CC orders move through the process correctly (getting either an approved/denied status)
 - Both actions leave the order on hold.
 - After an order gets a Signifyd status of approve or deny the order can be moved from hold without issue.
 - PayPal orders _always_ have a status of N/A.  Even though they are given a Signifyd score, they never leave N/A

I believe it is this extension causing the issue for two major reasons:
1. When the extension is disabled the issue does not exist.
2. When the extension is enabled and I attempt to unhold a paypal order, there is an entry in the file var/log/signifyd_connect.log that reads:

> [2017-10-20 15:13:36] SignifydLogger.INFO: Purchase Observer Order Hold: No: 10000XXXXX [] []

This line reads for the order number that was attempted to be changed.


Since this issue is preventing us from fulfilling orders I have also reached out to Signifyd support and discussed with Nicole on that side.  She informed us that:

>  There is a filter in place to prevent your PayPal orders from being submitted for guarantee with Signifyd, so by default the order status for your PayPal orders will show as HOLD. 

With a recommended fix of:

> With Magento 2.19 and above you can add which payment methods you would like to be excluded to the list on line 43 of file magento2/Observer/Purchase.php

OR

> edit the /vendor/signifyd/module-connect/Observer/purchase.php file.  Specifically line 43 to include paypal


Unfortunately, we are running Magento on their cloud platform.  We are unable to edit anything in the Magento Core or items in the /vendor folder.  Since the extension was installed via composer it is built during the Magento Cloud deploy process.


If the only way to fix this issue is to add paypal to the list of $restrictedMethods it would make sense to have this setting be available via the admin section.

Is there anyway to change this list through a custom extension on our end?  Maybe something that inherits from the existing class in the extension and updates that variable?  In this way it maintain all of the updates/compile steps.

I still think it would make sense to allow an edit in the admin section, but we need this resolved ASAP to fulfill these orders.  Our only other choice would be to discontinue use of Signifyd and we are hoping to avoid that.

Thank you,
James

**Comment by ebanolopes** [comment]
Hi, @James-Moon-Audio , that restricted list it is only for the development level and we recommend any changes on it in cases of having any issues like it is your case. That's why it stays only on the extension source code and it is not available on backend configurations.

I'll perform some more tests here with the details you've gave to me. Thanks for the detailed informations.

While I perform the tests, if you can provide to me:
1. Which PayPal solution you're using?
2. The PayPal configuration, because that can interfere on the results. If you can attach a print screen, that would be awesome!
3. Please attach here the signifyd_connect.log

To do not stop those orders from been fulfilled have you tried to:
a. Turn off the Signifyd functionality on configurations
b. Unhold the orders
c. Turn back on the Signifyd functionality

That would be a way to workaround this situation while we found a fix for it.

If you want to go on code level and try, as you mentioned, to modify the behavior of the extension using a custom extension, you can rewrite the \Signifyd\Connect\Observer\Purchase class using the "Class preference" method described on this link: http://inchoo.net/magento-2/overriding-classes-magento-2/
The only thing that your new class needs to have it is the $restrictedMethods property, it would be something like this:

```
<?php

namespace Vendor\Extension\Observer;

class Purchase extends \Signifyd\Connect\Observer\Purchase {
    //Add to bellow array the methods you want to restrict, do not remove any one
    protected $restrictedMethods = ['checkmo', 'banktransfer', 'purchaseorder', 'cashondelivery'];
}
```

If is there anything else I can do for you at this moment, please, let me know.

**Comment by James-Moon-Audio** [comment]
>    Which PayPal solution you're using?

PayPal All-in-One Payment Solutions
Payments Standard

>    The PayPal configuration, because that can interfere on the results. If you can attach a print screen, that would be awesome!

![screenshot1](https://user-images.githubusercontent.com/32907932/31836913-eb1828fe-b5a4-11e7-86e5-13a5dad1bb5b.jpg)
![screenshot2](https://user-images.githubusercontent.com/32907932/31836914-eb239d2e-b5a4-11e7-8002-1d56f6458075.jpg)

>    Please attach here the signifyd_connect.log

Here are the relevant lines:
[2017-10-20 15:02:41] SignifydLogger.INFO: Starting retry job [] []
[2017-10-20 15:02:41] SignifydLogger.INFO: Main retry method called [] []
[2017-10-20 15:02:41] SignifydLogger.INFO: Main retry method ended [] []
[2017-10-20 15:07:42] SignifydLogger.INFO: paypal_express [] []
[2017-10-20 15:07:42] SignifydLogger.INFO: Purchase Observer Order Hold: No: 1000029514 [] []
[2017-10-20 15:07:43] SignifydLogger.INFO: paypal_express [] []
[2017-10-20 15:12:55] SignifydLogger.INFO: paypal_express [] []
[2017-10-20 15:12:55] SignifydLogger.INFO: Purchase Observer Order Hold: No: 1000029514 [] []
[2017-10-20 15:12:55] SignifydLogger.INFO: paypal_express [] []
[2017-10-20 15:13:36] SignifydLogger.INFO: paypal_express [] []
[2017-10-20 15:13:36] SignifydLogger.INFO: Purchase Observer Order Hold: No: 1000029514 [] []
[2017-10-20 15:13:36] SignifydLogger.INFO: paypal_express [] []
[2017-10-20 15:22:45] SignifydLogger.INFO: braintree [] []
[2017-10-20 15:23:04] SignifydLogger.INFO: braintree [] []
[2017-10-20 15:23:40] SignifydLogger.INFO: Starting retry job [] []
[2017-10-20 15:23:40] SignifydLogger.INFO: Main retry method called [] []
[2017-10-20 15:23:40] SignifydLogger.INFO: Main retry method ended [] []

There are a lot of other entries for API requests, case saved, etc, but these seem to all be related to CC orders that are functioning as expected.

Your override class may be our best option.  I'm confused why this issue hasn't affected other Signifyd customers, do most other customers not filter out paypal orders from the Sygnifyd guarantee?

**Comment by ebanolopes** [comment]
Thanks for the informations fast provided @James-Moon-Audio 

About the issue, definitely it is not happening for all Signifyd customers using PayPal. Sometimes the issues could be more specific, like the store configuration, the payment method configuration and many other things that can cause different unexpected behaviors. I'll take a look at the information you provided and check if I'm able to reproduce the issue on my environment.

**Comment by ebanolopes** [comment]
@James-Moon-Audio I've performed some tests, and the orders didn't stays on the N/A status on any situation on my environment. Maybe there is a error on the webhooks configuration on Signifyd console or you have a issue with you cron configurations. Please, keep in touch with the Signifyd support team, they can assist you on this.

**Comment by joeshelton-wagento** [comment]
We also started having this issue after upgrading to Magento 2.2, using the core Signifyd module and realizing it doesn't have scores, switching back to this module. We are also on Magento Cloud hosting.

**Comment by ebanolopes** [comment]
@joeshelton-wagento , can you please check webhook configurations on Signifyd end, please? There is a [Test] button there, click on it to be sure that webhook requests are getting to Magento. Let me know about this test results.

**Comment by joeshelton-wagento** [comment]
I have a confusing report. Currently, we have the first-party Magento Signifyd module installed. We made this change when orders started being stuck in the On Hold status. The webhook tests all failed just now, likely because they were configured to use this module which has a different webhook URL.  The webhooks might not have been working back when this module was installed, but I can't verify that.

**Comment by ebanolopes** [comment]
The thing is: you can't have both build in and Signifyd extension enabled at same time on Magento 2.2 EE. That's because Magento has used same controller (which defines the webhook URL). On newer version of Signifyd extension we are putting a warning on admin area about this.

Go to web wizard setup and disable built in extension. After that, configure Signifyd extension, configure the correct webhooks on Signifyd console and try it again, please.

Let me know the results so I can assist you.

**Comment by joeshelton-wagento** [comment]
We already did this. When the "can't unhold orders" issue occurred we had Magento_Signifyd disabled in config.php

Webhook URL: https://www.domain.com/signifyd/webhooks/index

**Comment by ebanolopes** [comment]
That's the correct URL. If you access it on browser, what happens? The expected behavior is to see the message "You have successfully reached the webhook endpoint".

**Comment by joeshelton-wagento** [comment]
Oh, it's just occurring to me that restricting orders to the on hold status before a guarantee decision is intended functionality. That sheds a new light on things. But, I'm still confused. We did not have the Guarantee Completion webhook set up at any time, only Case webhooks. Yet this issue just recently emerged. I understand why orders are stuck on hold now, but I'm confused as to why orders were not stuck on hold previously.

What is your suggested configuration for a merchant that doesn't requests guarantees? My client simply uses the score to make the decision to process an order.

**Comment by joeshelton-wagento** [comment]
Added the Guarantee Competition webhook, and orders are still remaining in the N/A status. I found the following error in our logs.

Integrity constraint violation: 1062 Duplicate entry '' for key 'PRIMARY', query was: INSERT INTO `signifyd_connect_case` (`signifyd_status`, `code`, `score`, `entries_text`, `updated`, `magento_status`) VALUES (?, ?, ?, ?, '2017-12-01 07:46:07', ?)

It seems the order increment is not being added to the insert statement for some reason.
